### PR TITLE
Add support for customizing the extension install registry URL

### LINKED ISSUE #3374: Changing npm registry for installing extentions
Lens ver: 5.0.2-latests.20210705.2

I find out that's current registry for installing all extentions is hardcoded:

> export async function attemptInstallByInfo({ name, version, requireConfirmation = false }: ExtensionInfo) {
>   const disposer = ExtensionInstallationStateStore.startPreInstall();
>   const registryUrl = new URLParse("https://registry.npmjs.com").set("pathname", name).toString();

```
{url: "https://registry.npmjs.com/@nevalla/kube-resource-map", err: Error: Client network socket disconnected before secure TLS connection was established
    at connR…}err: Error: Client network socket disconnected before secure TLS connection was established
    at connResetException (internal/errors.js:570:14)
    at TLSSocket.onConnectEnd (_tls_wrap.js:1361:19)
    at Object.onceWrapper (events.js:312:28)
    at TLSSocket.emit (events.js:228:7)
    at endReadableNT (_stream_readable.js:1185:12)
    at processTicksAndRejections (internal/process/task_queues.js:81:21)url: "https://registry.npmjs.com/@nevalla/kube-resource-map"__proto__: Object
Promise.catch (async)
attemptInstallByInfo @ extensions.tsx:304
extensions_installFromInput @ extensions.tsx:452
installFromInput @ extensions.tsx:524
Rb @ react-dom.production.min.js:52
Xb @ react-dom.production.min.js:52
Yb @ react-dom.production.min.js:53
Ze @ react-dom.production.min.js:100
se @ react-dom.production.min.js:101
(anonymous) @ react-dom.production.min.js:113
Jb @ react-dom.production.min.js:292
Nb @ react-dom.production.min.js:50
jd @ react-dom.production.min.js:105
yc @ react-dom.production.min.js:75
hd @ react-dom.production.min.js:74
exports.unstable_runWithPriority @ scheduler.production.min.js:18
gg @ react-dom.production.min.js:122
Hb @ react-dom.production.min.js:292
gd @ react-dom.production.min.js:73
```

And because of this I can't use this behind a firewall
I want that lens uses my ~/.npmrc configuration file with my local registry

### PR DESCRIPTION
**Pull request by Nokel81**
Signed-off-by: Sebastian Malton <sebastian@malton.name>

fixes #3374

One of the new notifications (for instance if you run `npm config set registry https://google.ca`

Screen Shot 2021-07-15 at 5 41 17 PM

### CODE REVIEW

File: src/renderer/components/+extensions/extensions.tsx [hunk 1]
```diff
@@ -300,16 +302,58 @@ async function unpackExtension(request: InstallRequestValidated, disposeDownload
...
+      try {
+        const filteredEnv = Object.fromEntries(
+          Object.entries(process.env)
+            .filter(([key]) => !key.startsWith("npm")),
+        );
+        const { stdout } = await promiseExecFile("npm", ["config", "get", "registry"], { env: filteredEnv });
+
+        return stdout.trim();
+      } catch (error) {
+        console.warn("[EXTENSIONS]: failed to get configured registry from .npmrc", error);
```
Comment (aleksfront) [inline src/renderer/components/+extensions/extensions.tsx]: Should we show extra notification for this scenario?

**Review by aleksfront — DISMISSED** [review]
LGTM

**Comment by aleksfront** [comment]
@Nokel81 I've added small UI changes - for always showing custom URL input field (and changing its `disabled` state)
registry setting

File: src/renderer/components/+extensions/extensions.tsx (hunk 1 above)
Comment (Nokel81) [reply src/renderer/components/+extensions/extensions.tsx]: Sure

**Review by aleksfront — DISMISSED** [review]

File: src/renderer/components/+preferences/application.tsx [hunk 2]
```diff
@@ -93,13 +97,47 @@ export const Application = observer(() => {
...
+            if (userStore.extensionRegistryUrl.location === ExtensionRegistryLocation.CUSTOM) {
+              userStore.extensionRegistryUrl.customUrl = "";
+            }
+          })}
+          themeName="lens"
+        />
+        <p className="mt-4 mb-5 leading-relaxed">
+          This setting is to change the registry URL for installing extensions by name.{" "}
+          If you are unable to access the default registry ({defaultExtensionRegistryUrl}) {" "}
+          you should change this to either the registry in you{" "}
```
Comment (jweak) [inline src/renderer/components/+preferences/application.tsx]: [typo] you => your

Maybe rephrase it `you can change the registry URL in your .npmrc or add a custom URL in the input below`. 

Or something like that.

File: src/renderer/components/+preferences/application.tsx [hunk 3]
```diff
@@ -93,13 +97,47 @@ export const Application = observer(() => {
 
       <hr/>
 
+      <section id="extensionRegitryUrl">
```
Comment (jweak) [inline src/renderer/components/+preferences/application.tsx]: typo
`extensionRegitryUrl` => `extensionRegistryUrl`

File: src/common/user-store/preferences-helpers.ts [hunk 4]
```diff
@@ -306,6 +306,36 @@ const updateChannel: PreferenceDescription<string> = {
...
+}
+export type ExtensionRegistry = {
+  location: ExtensionRegistryLocation.DEFAULT | ExtensionRegistryLocation.NPMRC;
+  customUrl?: undefined;
+} | {
+  location: ExtensionRegistryLocation.CUSTOM,
+  customUrl: string;
+};
+
+export const defaultExtensionRegistryUrl = "https://registry.npmjs.com";
```
Comment (jweak) [inline src/common/user-store/preferences-helpers.ts]: It says the [here]( https://docs.npmjs.com/cli/v8/using-npm/registry) that the default domain is .org

> npm is configured to use the npm public registry at https://registry.npmjs.org by default.

Should we use the same as default?

Comment (Nokel81) [reply src/common/user-store/preferences-helpers.ts]: Interesting, both work... maybe we should then

**Review by jweak — DISMISSED** [review]
LGTM

File: src/renderer/components/+extensions/get-base-registry-url/get-base-registry-url.injectable.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+import { Injectable, lifecycleEnum } from "@ogre-tools/injectable";
+import { UserStore } from "../../../../common/user-store";
+import { Dependencies, getBaseRegistryUrl } from "./get-base-registry-url";
+
+const getBaseRegistryUrlInjectable: Injectable<() => Promise<string>, Dependencies> = {
+  getDependencies: () => ({
+    // TODO: use injection
+    getRegistryUrlPreference: () => UserStore.getInstance().extensionRegistryUrl,
```
Comment (Iku-turso) [inline src/renderer/components/+extensions/get-base-registry-url/get-base-registry-url.injectable.ts:29]: Yes, let's agree to *not* use injection here yet, as making it happen for *Stores needs base work. We already have some eyes on this, but it's not happening in a couple of days.

File: src/renderer/components/+preferences/application.tsx [hunk 6]
```diff
@@ -49,7 +52,8 @@ export const Application = observer(() => {
         : "System default shell"
     );
 
-  const [shell, setShell] = React.useState(UserStore.getInstance().shell || "");
+  const [shell, setShell] = React.useState(userStore.shell || "");
```
Comment (Iku-turso) [inline src/renderer/components/+preferences/application.tsx]: In general, I have a bit of an issue with using React state and MobX state wantonly, but let's make a bigger deal of that somewhere else, some other time.

**Review by Iku-turso — DISMISSED** [review]

**Review by Iku-turso — APPROVED** [review]

**Review by jansav — APPROVED** [review]
